marlowe-playground-client/src/Halogen/BlocklyCommons.purs

```diff
@@ -0,0 +1,71 @@
+-- TODO In a different commit I'd rename Blockly to BlocklyEditor and ActusBlockly as ActusEditor or
+-- ActusBlocklyEditor and this module could be called solely Blockly
+module Halogen.BlocklyCommons where
+
+import Blockly (addChangeListener, removeChangeListener)
+import Blockly.Events (fromEvent, newParentId, oldParentId)
+import Blockly.Types (BlocklyEvent, Workspace)
+import Blockly.Types as BT
+import Control.Alt ((<|>))
+import Data.Foldable (oneOf)
+import Data.Lens (Lens', assign)
+import Data.Lens.Record (prop)
+import Data.Symbol (SProxy(..))
+import Data.Traversable (for_)
+import Effect.Aff.Class (class MonadAff)
+import Halogen (HalogenM, raise)
+import Halogen.Query.EventSource (EventSource)
+import Halogen.Query.EventSource as EventSource
+import Prelude (Unit, bind, discard, pure, unit, ($), (<$>))
+import Web.Event.EventTarget (eventListener)
+
+blocklyEvents ::
+  forall m action.
+  MonadAff m =>
+  (BlocklyEvent -> action) ->
+  Workspace ->
+  EventSource m action
+blocklyEvents toAction workspace =
+  EventSource.effectEventSource \emitter -> do
+    listener <-
+      eventListener \event -> do
+        let
+          mEvent =
+            -- Blockly can fire all of the following events https://developers.google.com/blockly/guides/configure/web/events
+            -- but at the moment we only care for the ones that may affect the unsaved changes.
+            oneOf
+              [ BT.Create <$> fromEvent event
+              , BT.Move <$> fromEvent event
+              , BT.Change <$> fromEvent event
+              , BT.FinishLoading <$> fromEvent event
+              ]
+        for_ mEvent \ev -> EventSource.emit emitter (toAction ev)
+    addChangeListener workspace listener
+    pure $ EventSource.Finalizer $ removeChangeListener workspace listener
+
+_hasUnsavedChanges' :: forall r. Lens' { hasUnsavedChanges :: Boolean | r } Boolean
+_hasUnsavedChanges' = prop (SProxy :: SProxy "hasUnsavedChanges")
+
+updateUnsavedChangesActionHandler ::
+  forall m state action slots message.
+  MonadAff m =>
+  message ->
+  message ->
+  BlocklyEvent ->
+  HalogenM { hasUnsavedChanges :: Boolean | state } action slots message m Unit
+updateUnsavedChangesActionHandler codeChange finishLoading event = do
+  let
+    setUnsavedChangesToTrue = do
+      assign _hasUnsavedChanges' true
+      raise codeChange
+  case event of
+    (BT.Change _) -> setUnsavedChangesToTrue
+    -- The move event only changes the unsaved status if the parent has changed (either by attaching or detaching
+    -- one block into another)
+    (BT.Move ev) -> for_ (newParentId ev <|> oldParentId ev) \_ -> setUnsavedChangesToTrue
```

Comment:
I am not sure I understand this, if the parent has changed I would imagine the condition would be something like `newParent /= oldParent`. I am guessing this checks whether at least one of the sides of the `<|>` is `Just`. I may be completely off

Reply:
Your guess is right, it just checks that newParentId or oldParentId is defined. 
When you attach a component to another, then newParentId is defined but oldParentId probably not, and when you detach a block from it's parent oldParentId is defined but newParentId will not

Reply:
Then I think it would be more robust to check for `/=`, cause it covers that case and also the potential case where you change from one parent to another (which may be impossible but you never know). And it is also simpler and easier to understand, I think it would be something like:
```haskell
BT.Move ev -> when (newParentId ev /= oldParentId ev) setUnsavedChangesToTrue
```
Now that I think about it you probably don't need parenthesis around `BT.Move ev` and `BT.Change _`... either

Reply:
Yeah, it's clearer that way :)

done